wpilibc/src/test/native/cpp/event/BooleanEventTest.cpp

```diff
@@ -0,0 +1,486 @@
+
+  (BooleanEvent(&loop, [] { return true; }) && BooleanEvent(&loop, [] {
+     return false;
+   })).IfHigh([&] { ++andCounter; });
+  (BooleanEvent(&loop, [] { return true; }) || BooleanEvent(&loop, [] {
+     return false;
+   })).IfHigh([&] { ++orCounter; });
+
+  loop.Poll();
+
+  EXPECT_EQ(0, andCounter);
+  EXPECT_EQ(1, orCounter);
+}
+
+/**
+ * Tests that composed edge events only execute on edges (two
+ * rising edge events composed with and() should only execute when both signals
+ * are on the rising edge)
+ */
+TEST(BooleanEventTest, BinaryCompositionsWithEdgeDecorators) {
+  EventLoop loop;
+  bool boolean1 = false;
+  bool boolean2 = false;
+  bool boolean3 = false;
+  bool boolean4 = false;
+  int counter = 0;
+
+  auto event1 = BooleanEvent(&loop, [&] { return boolean1; }).Rising();
+  auto event2 = BooleanEvent(&loop, [&] { return boolean2; }).Rising();
+  auto event3 = BooleanEvent(&loop, [&] { return boolean3; }).Rising();
+  auto event4 = BooleanEvent(&loop, [&] { return boolean4; }).Rising();
+  (event1 && event2).IfHigh([&] { ++counter; });
+  (event3 || event4).IfHigh([&] { ++counter; });
+  EXPECT_EQ(0, counter);
+
+  boolean1 = true;
+  boolean2 = true;
+  boolean3 = true;
+  boolean4 = true;
+  loop.Poll();  // Both actions execute
+
+  EXPECT_EQ(2, counter);
+
+  loop.Poll();  // Nothing should happen since nothing is on rising edge
+
+  EXPECT_EQ(2, counter);
+
+  boolean1 = false;
+  boolean2 = false;
+  boolean3 = false;
+  boolean4 = false;
+  loop.Poll();  // Nothing should happen
+
+  EXPECT_EQ(2, counter);
+
+  boolean1 = true;
+  loop.Poll();  // Nothing should happen since only Bool 1 is on rising edge
+
+  EXPECT_EQ(2, counter);
+
+  boolean2 = true;
+  loop.Poll();  // Bool 2 is on rising edge, but Bool 1 isn't, nothing should
+                // happen
+
+  EXPECT_EQ(2, counter);
+
+  boolean1 = false;
+  boolean2 = false;
+  loop.Poll();  // Nothing should happen
+
+  EXPECT_EQ(2, counter);
+
+  boolean1 = true;
+  boolean2 = true;
+  loop.Poll();  // Bool 1 and 2 are on rising edge, increments counter once
+
+  EXPECT_EQ(3, counter);
+
+  boolean3 = true;
+  loop.Poll();  // Bool 3 is on rising edge, increments counter once
+
+  EXPECT_EQ(4, counter);
+
+  loop.Poll();  // Nothing should happen, Bool 3 isn't on rising edge
+
+  EXPECT_EQ(4, counter);
+
+  boolean4 = true;
+  loop.Poll();  // Bool 4 is on rising edge, increments counter once
+
+  EXPECT_EQ(5, counter);
+
+  loop.Poll();  // Nothing should happen, Bool 4 isn't on rising edge
+
+  EXPECT_EQ(5, counter);
+}
+
+TEST(BooleanEventTest, BinaryCompositionLoopSemantics) {
+  EventLoop loop1;
+  EventLoop loop2;
+  bool boolean1 = true;
+  bool boolean2 = true;
+  int counter1 = 0;
+  int counter2 = 0;
+
+  (BooleanEvent(&loop1, [&] { return boolean1; }) && BooleanEvent(&loop2, [&] {
+     return boolean2;
+   })).IfHigh([&] { ++counter1; });
+  (BooleanEvent(&loop2, [&] { return boolean1; }) && BooleanEvent(&loop1, [&] {
+     return boolean2;
```

Comment:
Is the cross-use here (loop2 updating boolean1 and vice versa) important to the test? It adds a lot of complexity.
Each loop reading from the similarly-indexed boolean and updating the similarly-indexed counter would be much simpler to track. (ie the second event composition would be identical to the first, except the operands are flipped)
Would the test still fulfill its purpose then?